distributed/scheduler.py

```diff
@@ -6250,6 +6288,10 @@ async def retire_workers(
                         lock=False,
                     )
 
+                for group in groups:
+                    if group._last_worker in workers:
+                        group._last_worker = None
```

Comment:
This seems like an overly specific place to put this logic.  If the task goes out of scope for any other reason then it won't hit this area.

Instead, maybe we can clear the last worker when we find that there are no more workers in the released or waiting state?  I think that this is what I did in my PR and it seemed to work out ok.  (tests passed at least)

Reply:
Since the goal of all this is just to avoid `return last` returning a worker that doesn't exist anymore, why don't we check `if last.address in self._workers_dv`? When we were in the standalone `decide_worker`, we only had the dict_values, so that check would have been O(n), but now that we can do the O(1) lookup on the dict, that seems like the simplest solution.

Reply:
That sounds great to me.  

Thanks for keeping this change isolated.  That feels like a good change to me.